Document udev config to run evsieve

azuwis opened the pull request.

azuwis committed:
Document udev config to run evsieve

KarsMulder commented:
Hi! Thanks for your contribution. I haven't had the time to dive into the underlying udev and systemd mechanics yet (and I feel like I shouldn't merge things I don't understand, so that's why the pull request is still on hold), but I wanted to post a reply anyway to tell you that this pull request hasn't been forgotten.

Some thoughts so far...

```
[Service]
Type=simple
DynamicUser=yes
SupplementaryGroups=input uinput
```

This is probably different on your operating system, but on my operating system (Arch Linux), there is no uinput group, and worse, there is no non-root group that has write permissions by default for `/dev/uinput`:

```
$ ls -l /dev/uinput
crw-rw----+ 1 root root 10, 223 Nov 15 23:21 /dev/uinput

$ getfacl /dev/uinput 
getfacl: Removing leading '/' from absolute path names
# file: dev/uinput
# owner: root
# group: root
user::rw-
user:brltty:rw-
group::---
mask::rw-
other::---
```

So the `DynamicUser` approach is unfortunately non-portable.

(The brltty group was made by a package that is not part of the base system, so its existence cannot be relied on.)

---

Next point:

```
[Service]
Type=simple
```

This part isn't wrong, but since evsieve explicitly supports systemd's notify protocol, I feel like this should be `Type=notify` to be consistent with the other parts of the documentation. This use case doesn't actually need the notify protocol, but I think the user should not be confused "wait, should I use `Type=simple` or `Type=notify` when I make a systemd service?"

... except `Type=notify` is known to be broken when evsieve is ran as non-root unless `NotifyAccess=all` is also set, so that line would have to be added too if continuing with the `DynamicUser` approach.

(I kinda hoped that that limitation would never actually matter, but here we are...)

---

Final point: currently, evsieve does indeed exit when the input device is closed. That behaviour was chosen because of backward compatibility reasons. I wanted to change that default behaviour to `persist=reopen` in a hypothetical evsieve 2.0. Since this configuration explicitly relies on evsieve exiting when the input device disappears, I feel like I should explicitly specify `persist=none` or `persist=exit` to avoid recommending scripts that might break in a future version.

(Relatedly, thanks for the good example of valid use cases that will actually break if that default option is changed. I'm starting to reconsider whether that particular option should change even if I were to ever release evsieve 2.0.)

azuwis committed:
Update udev doc

- Step to add uinput group
- Udev rule to make `/dev/uinput` writable by `uinput` group
- Change to use `sudo udevadm trigger` for the uinput udev rule
- Change systemd service type to notify
- Set persist=exit for evsieve command line args

KarsMulder commented:
I've decided to merge this pull request, but made some changes to it (commit ebd7efe1ee902e70c5943b65a2bf44b9a3c31eb8).

The idea of using udev to start evsieve, and the way it combines with systemd template services, is great. Thank you for this contribution!

I've removed the part that recommends anything involving systemd's user mode. As you can read in the commit message, I think that recommending giving the user's own account to read/write to the input devices and `/dev/uinput` is not appropriate from a security perspective. If a user belongs to the `input` group, any program the run can keylog events from all other programs, and even programs from other user if there are any. That was pretty much the status quo under X11, but Wayland has been trying to patch it. You may decide that such attacks do not fall under your threat model, but it should not recommended nilly-willy to users who do not understand the consequences of doing so.

The part I was more ambivalent about was the part involving DynamicUser. From a security perspective, it is certainly an improvement compared to running evsieve as root. However, the threats that it protects against are mostly theoretic. The attack surface of evsieve is mainly its input handling. Even if there was a security hole in evsieve and there existed a certain particular sequence of input events would cause evsieve to misbehave and permit privilege escalation, how would an attacker force that particular sequence of events to appear without already having very high privileges?

The additional sequence of steps required to set this up makes an already complicated thing even more complicated, increases the risk that the user screws something up, and generally feels like it messes with things that only the operating system should mess with, e.g. what if a certain OS decided to make the `/dev/uinput` file owned by `root:input`, but the user just follows the instructions caused the ownership to be changed to `root:uinput` and subsequently breaks his operating system?

If some kind of specialized operating system were to ship prewritten evsieve configuration files to the end user, then it may indeed be worth the maintainers' time to setup a systemd configuration involving DynamicUser, but I think it is overkill to recommend it to the average end-user, that's why I removed that part as well.
